src/test/java/uk/gov/ons/ctp/integration/envoycuc/mockclient/MockClient.java

```diff
@@ -0,0 +1,292 @@
+
+  @PostConstruct
+  private void clearMaps() {
+    allowanceMap.clear();
+    postingsTimeMap.clear();
+  }
+
+  public RateLimitResponse checkRateLimit(
+      RateLimiterClient.Domain domain,
+      Product product,
+      CaseType caseType,
+      String ipAddress,
+      UniquePropertyReferenceNumber uprn,
+      String telNo) {
+
+    final RateLimiterClientRequest rateLimiterClientRequest =
+        new RateLimiterClientRequest(domain, product, caseType, ipAddress, uprn, telNo);
+    final IsValidRequest isValidRequest = postRequest(rateLimiterClientRequest);
+
+    if (!isValidRequest.isValid()) { // invalid request - throw a 429
+      final StringBuilder reason = new StringBuilder("Too Many Requests - ");
+      isValidRequest
+          .getLimitStatusList()
+          .forEach(stat -> reason.append(stat.toString()).append(" - "));
+      throw new ResponseStatusException(
+          org.springframework.http.HttpStatus.TOO_MANY_REQUESTS, reason.toString());
+    }
+
+    RateLimitResponse response = new RateLimitResponse();
+    response.setOverallCode("200");
+    response.setStatuses(isValidRequest.getLimitStatusList());
+    final StringBuilder reason = new StringBuilder("Request Successful - ");
+    isValidRequest
+        .getLimitStatusList()
+        .forEach(stat -> reason.append(stat.toString()).append(" - "));
+    log.info(reason.toString());
+    return response;
+  }
+
+  public IsValidRequest postRequest(final RateLimiterClientRequest rateLimiterClientRequest)
+      throws ResponseStatusException {
+
+    List<String> requestKeyList = getKeys(rateLimiterClientRequest);
+    final IsValidRequest isValidRequest =
+        isValidateRequest(requestKeyList, rateLimiterClientRequest);
+    postRequest(
+        requestKeyList,
+        rateLimiterClientRequest); // always post - it burns allowances every time for all scenarios
+
+    return isValidRequest;
+  }
+
+  private List<String> getKeys(RateLimiterClientRequest request) {
+    List<String> keyList = new ArrayList<>();
+    StringBuilder keyBuff =
+        new StringBuilder(request.getProduct().getProductGroup().name().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getIndividual().toString().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getDeliveryChannel().name().toUpperCase())
+            .append("-")
+            .append(request.getCaseType().name().toUpperCase())
+            .append("-");
+    if (request.getUprn() != null && request.getUprn().getValue() != 0) {
+      keyList.add(keyBuff.toString() + "UPRN");
+    }
+    if (request.getIpAddress() != null) {
+      keyList.add(keyBuff.toString() + "IP");
+    }
+    if (request.getTelNo() != null) {
+      keyList.add(keyBuff.toString() + "TELNO");
+    }
+    return keyList;
+  }
+
+  private IsValidRequest isValidateRequest(
+      List<String> requestKeyList, RateLimiterClientRequest f) {
```

Comment:
"f" , might be easier to read as "request" as used elsewhere.

Reply:
It was called something else that had a wopper name fulfilmentcontext blah blah etc, but it's not relevant now

clientRequest